jupyterhub/scopes.py

```diff
@@ -6,33 +6,36 @@
 from tornado.log import app_log
 
 from . import orm
+from . import roles
 
 
 class Scope(Enum):
     ALL = True
 
 
-def get_user_scopes(name):
-    """
-    Scopes have a metascope 'all' that should be expanded to everything a user can do.
-    At the moment that is a user-filtered version (optional read) access to
-    users
-    users:name
-    users:groups
-    users:activity
-    users:servers
-    users:tokens
-    """
-    scope_list = [
-        'users',
-        'users:name',
-        'users:groups',
-        'users:activity',
-        'users:servers',
-        'users:tokens',
-    ]
-    scope_list.extend(['read:' + scope for scope in scope_list])
-    return {"{}!user={}".format(scope, name) for scope in scope_list}
+def get_scopes_for(orm_object):
+    """Find scopes for a given user or token"""
+    scopes = set()
+    if orm_object is None:
+        return scopes
+    elif isinstance(orm_object, orm.APIToken):
+        owner = orm_object.user or orm_object.service
+        token_scopes = roles.get_scopes_for(orm_object)
+        owner_scopes = roles.get_scopes_for(owner)
+        if 'all' in token_scopes:
+            token_scopes.remove('all')
+            token_scopes |= owner_scopes
+        scopes = token_scopes & owner_scopes
+        discarded_token_scopes = token_scopes - scopes
+        # Not taking symmetric difference here because token owner can naturally have more scopes than token
+        if discarded_token_scopes:
+            app_log.warning(
+                "Token-based access, discarding scopes [%s]"
+                % ", ".join(discarded_token_scopes)
```

Comment:
It's probably important to include the token object's repr itself in the logs, otherwise it will be hard to identify which token is responsible for the warning.